flask/app/gene_viewer.py

```diff
@@ -0,0 +1,248 @@
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters), models.users_groups_table.columns.user_id == user_id)
+        )
+
+    else:
+        app.logger.debug("Processing query - unrestricted based on user id.")
+        query = (
+            models.Gene.query.join(models.Gene.variants)
+            .join(models.Variant.genotype)
+            .join(models.Genotype.analysis, models.Genotype.dataset)
+            .join(models.Dataset.tissue_sample)
+            .join(models.TissueSample.participant)
+            .join(models.Participant.family)
+            .options(
+                contains_eager(models.Gene.variants)
+                .contains_eager(models.Variant.genotype)
+                .contains_eager(models.Genotype.analysis)
+                .contains_eager(models.Analysis.datasets)
+                .contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters))
+        )
+
+    q = query.all()
+
+    if len(q) == 0:
+        app.logger.error("No requested genes were found.")
+        abort(400, description="No requested genes were found.")
+    elif len(q) < len(genes):
+        app.logger.error("Not all requested genes were found.")
+        abort(400, description="Not all requested genes were found.")
+
+    if request.accept_mimetypes["application/json"]:
+
+        app.logger.info("Application/json Accept header requested")
+        return jsonify(
+            [
+                {
+                    **asdict(gene),
+                    "variants": [
+                        {
+                            **asdict(variant),
+                            "zygosity": [
+                                genotype.zygosity for genotype in variant.genotype
+                            ],
+                            "alt_depths": [
+                                genotype.alt_depths for genotype in variant.genotype
+                            ],
+                            "burden": [
+                                genotype.burden for genotype in variant.genotype
+                            ],
+                            "participant_codenames": [
+                                dataset.tissue_sample.participant.participant_codename
+                                for dataset in variant.analysis.datasets
+                            ],
+                        }
+                        for variant in gene.variants
+                    ],
+                }
+                for gene in q
+            ]
+        )
+
+    elif request.accept_mimetypes["text/csv"]:
+
+        app.logger.info("text/csv Accept header requested")
+
+        try:
+            agg_df = get_variant_wise_df(query.statement, query.session.bind)
```

Comment:
Is pandas performing the query again after `query.all()` already fetches everything?